### DISCUSSION TITLE
How to map split strings from file into fields with jq?

### DISCUSSION THREAD

**Discussion by drankinatty** [stedolan/jq#3197] [Q&A] [answered]
This seems simple (and it is for those skilled with `jq`), but my experience is lacking. I have a simple file that contains words and definitions, one per-line, pipe (`'|'`) separated, e.g.:

```none
$ head -n1 words.txt
Appreciable|describes things that can be perceived or measured.
```

I am trying to read the file into a json document with the normal array of arrays structure but giving the words and definitions tags of `word:` and `defn:`. The structure will be:

```
[
  [
    "word": "Appreciable",
    "defn": "describes things that can be perceived or measured."
  ]
]
```

Testing with `input` to work out the filter, I can get the data in the form needed, but I cannot figure out how to add the tags. For example using `split ("|")` produces the document I want:

```none
$ jq -R -n '[input | split("|")]' words.txt
[
  [
    "Appreciable",
    "describes things that can be perceived or measured."
  ]
]
```

But when I try to `map()` the fields, I end up with errors of `Cannot index string with number` or the like.  I can't sort out how I need to continue the pipeline to `... | map ({word: <what here?>}, {defn: <what here?>})` to complete the document. I've gone through the 1.7 manual and I cannot find how to do what I'm trying to do. (that tells me I don't know enough to recognize what I'm looking at)

I suspect I need to take the result of `split()` and create an array with it (something like `as [$word, $defn]` and then be able to complete the map with the variables -- but the implementation escapes me.

How do I simply add the `"word":` and `"defn":` tags to the split strings?

**Comment by wader** [comment, marked as answer]
Hi, your close! maybe something like this?
```sh
$ echo -e 'a|b\nc|d' | jq -nR '[inputs | split("|") | {word: .[0], defn: .[1]}]'
[
  {
    "word": "a",
    "defn": "b"
  },
  {
    "word": "c",
    "defn": "d"
  }
]
```

**Reply by drankinatty** [reply]
Aaarrrgggghhh! Brilliant. I didn't need `map()` at all, it was just providing the tags!

Thank you! It's funny, I'm quite adept with `sed`, `awk`, POSIX shell, C, C++, FORTRAN, assembly, etc.., but `jq` is something I don't use enough to have made friends with it yet. I'd spent an hour or so going though posts on StackOverflow, Unix & LInux, and the wider web and pawing through the `jq` manual -- and I had it stuck in my head that `map()` was the key. One certain way for humans to waste time is to start off down the wrong path.... It has the momentum of an aircraft carrier at that point.

Thank you for getting me back on the proper course.

**Reply by wader** [reply]
👍 it's a bit special with its tacit syntax and generators, i can recommend having a look at https://github.com/jqlang/jq/blob/master/src/builtin.jq, that was enlightening for me :)

**Comment by pkoppstein** [comment]
@drankinatty - Nice question.  I hope you'll find your jq explorations worthwhile.  For future reference, please ask usage questions at stackoverflow.com, not least because that will likely help others find good questions and helpful answers.

**Reply by drankinatty** [reply]
I wondered about that. I've contributed on SO for a decade and the only reason I asked here, was my search turned up [Simplest way to add content of the text file as json string field using jq](https://github.com/jqlang/jq/discussions/2908) which led me to ask here. Will post all additional usage questions (hopefully few after I wrap my head around the pipeline order and flow), on StackOverflow. Sorry for the noise.

**Reply by wader** [reply]
I used to contribute a lot to SO also but felt it got so competitive and hostile at times so i stopped :( so personally i would not mind if the jq discussions was used more and we built a knowledge base here.
